### PR TITLE
add attributes to some unittests of std.range 

### PR DESCRIPTION
**Pull request by dukc**
Added attributes to some std.range.package unittests so they have a better chance to catch regressions.

### CODE REVIEW

File: std/range/package.d [hunk 1]
```diff
@@ -317,16 +317,17 @@ if (isBidirectionalRange!(Unqual!Range))
...
-@safe unittest
+pure @safe nothrow @nogc unittest
 {
     import std.algorithm.comparison : equal;
-    int[] a = [ 1, 2, 3, 4, 5 ];
-    assert(equal(retro(a), [ 5, 4, 3, 2, 1 ][]));
-    assert(retro(a).source is a);
-    assert(retro(retro(a)) is a);
+    int[5] a = [ 1, 2, 3, 4, 5 ];
+    int[5] b = [ 5, 4, 3, 2, 1 ];
```
Comment (wilzbach) [inline std/range/package.d]: You might be interested in this PR for automatically inferring the size of static arrays: https://github.com/dlang/phobos/pull/4936

File: std/range/package.d [hunk 2]
```diff
@@ -317,16 +317,17 @@ if (isBidirectionalRange!(Unqual!Range))
 
 
 ///
-@safe unittest
+pure @safe nothrow @nogc unittest
```
Comment (wilzbach) [inline std/range/package.d]: Is there already a [D-style](http://dlang.org/dstyle.html) way to order for attributes?
Would probably make sense to do so and thus have a common style in Phobos.

File: std/range/package.d [hunk 3]
```diff
@@ -646,22 +648,31 @@ unittest
...
     int[4] testArr = [1,2,3,4];
     //just checking it compiles
     auto s = testArr[].stride(2);
 }
 
-debug unittest
+debug pure nothrow unittest
 {//check the contract
     int[4] testArr = [1,2,3,4];
+    bool passed = false;
```
Comment (wilzbach) [inline std/range/package.d]: how about `scope(exit) assert(passed)`?

File: std/range/package.d [hunk 4]
```diff
@@ -2419,15 +2441,15 @@ if (isInputRange!R)
...
+pure @safe nothrow unittest
 {
     import std.algorithm.comparison : equal;
     import std.internal.test.dummyrange : DummyRange, Length, RangeType, ReturnBy;
 
     alias DummyType = DummyRange!(ReturnBy.Value, Length.No, RangeType.Forward);
     auto te = takeExactly(DummyType(), 5);
     Take!DummyType t = te;
-    assert(equal(t, [1, 2, 3, 4, 5]));
+    assert(equal(t, [1, 2, 3, 4, 5][]));
```
Comment (wilzbach) [inline std/range/package.d]: Why did you add the extra `[]` slicing here?

**Comment by wilzbach** [comment]
> We need more enforcements like this. Thanks!

There are already some in place - see the failing CircleCI message:

```
Enforce whitespace before opening parenthesis
grep -nrE "(for|foreach|foreach_reverse|if|while|switch|catch)\(" $(find . -name '*.d') ; test $? -eq 1
./std/range/package.d:668:    catch(AssertError unused)
./std/range/package.d:2159:    catch(UTFException unused)
./std/range/package.d:2163:    catch(Exception unused)
./std/range/package.d:2823:    catch(UTFException unused)
./std/range/package.d:2827:    catch(Exception unused)
./std/range/package.d:3049:    catch(UTFException unused)
./std/range/package.d:3053:    catch(Exception unused)
```

(you can run the style tests locally with `make -f posix.mak style`).

**Comment by dukc** [comment]
starting to fix now...

File: std/range/package.d (hunk 1 above)
Comment (dukc) [inline std/range/package.d]: or I could make them static immutable int[], as I saw done later.

**Comment by dukc** [comment]
how I test for style at windows? make -f win[32/64].mak style compains:
Error: don't know how to make 'style'

**Comment by wilzbach** [comment]
> how I test for style at windows? make -f win[32/64].mak style complains:
> Error: don't know how to make 'style'

At the moment it's Posix only:
https://github.com/dlang/phobos/blob/master/posix.mak#L497

However, I am [trying](https://github.com/Hackerpilot/dscanner/pulls) to move the "simple grep" style checks over to Dscanner, s.t. it's truly cross-platform, but this is pending since quite a while.

**Comment by dukc** [comment]
What does that codecov/patch test failure mean I did wrong? After all, I didn't add any new code outside unittests so I don't see how I worsened unittest coverage ratio, if it means that.

**Comment by wilzbach** [comment]
> What does that codecov/patch test failure mean I did wrong? After all, I didn't add any new code outside unittests so I don't see how I worsened unittest coverage ratio, if it means that.

If you click on the link or [install the browser extension](https://github.com/codecov/browser-extension) you will see the reason:

image

See [this discussion](http://forum.dlang.org/post/nncm00$2sc8$1@digitalmars.com) for an explanation.
tl;dr: the DMD compiler doesn't ignore `unittest` blocks and it won't do so in the foreseeable future.

**Comment by dukc** [comment]
Thanks, I see. Trough, I'm not sure how I should act in blocks like this...

File: std/range/package.d [hunk 5]
```diff
@@ -646,22 +648,31 @@ unittest
...
-debug unittest
+debug pure nothrow unittest
 {//check the contract
     int[4] testArr = [1,2,3,4];
+    bool passed = false;
+    scope (success) assert(passed);
     import core.exception : AssertError;
-    import std.exception : assertThrown;
-    assertThrown!AssertError(testArr[].stride(0));
+    //std.exception.assertThrown won't do because it can't infer nothrow
```
Comment (wilzbach) [inline std/range/package.d]: this is due to https://issues.dlang.org/show_bug.cgi?id=12647

Comment (dukc) [inline std/range/package.d]: Yes, but should I do something for it? Only things coming to mind are unittesting the unittest or discarding the whole idea of making it nothrow.

Comment (dukc) [inline std/range/package.d]: ...or, I could try making a template where the experession is taken as an alias parameter, but I think that is so general it should be another pr.

Comment (wilzbach) [inline std/range/package.d]: > Yes, but should I do something for it?

Nope, not much you can do - though there's a fix in progress: https://github.com/dlang/dmd/pull/6348

**Comment by wilzbach** [comment]
Hmm one never knows how long a PR will take to land in master, so I suggest to mark the "ugly" `try /catch` bits with a  `// @@@BUG@@@ 12647` comment, s.t. they are easy to find once the fix for the bug is there.
-> afterwards squash your commits into one & we should be good to go :)

**Comment by dukc** [comment]
Thank you very much, again. I probably have to exercise first with a branch I've not made a pr yet of.

**Comment by edi33416** [comment]
@dukc We have a pretty good getting started guide. It also gives instructions on how to [rebase](https://wiki.dlang.org/Starting_as_a_Contributor#Rebasing)

Nice job with the PR

**Comment by wilzbach** [comment]
> Nice job with the PR

Looking forward to more!
